lib/src/resolve-symlinks.js

```diff
@@ -1,52 +1,44 @@
 'use strict';
 
-var through2 = require('through2');
+var through = require('through2');
 var fs = require('graceful-fs');
 
 var koalas = require('koalas');
 var valueOrFunction = require('value-or-function');
 var boolean = valueOrFunction.boolean;
 
-function resolveSymlinks(options) {
+function resolveSymlinks(opt) {
 
-  var resolveSymlinks = koalas(boolean(options.resolveSymlinks), true);
+  var resolveSymlinks = koalas(boolean(opt.resolveSymlinks), true);
 
   // A stat property is exposed on file objects as a (wanted) side effect
   function resolveFile(globFile, enc, callback) {
+    var stat = opt.statCache[globFile.path];
+    if (stat && resolveSymlinks) {
+      // No need for additional filesystem calls
+      globFile.stat = stat;
+      return callback(null, globFile);
+    }
+
     fs.lstat(globFile.path, onStat);
 
     function onStat(statErr, stat) {
       if (statErr) {
         return callback(statErr);
       }
 
-      globFile.stat = stat;
-
       if (!stat.isSymbolicLink() || !resolveSymlinks) {
+        opt.statCache[globFile.path] = stat;
+        globFile.stat = stat;
         return callback(null, globFile);
       }
 
-      fs.realpath(globFile.path, onRealpath);
-    }
-
-    function onRealpath(realpathErr, filePath) {
-      if (realpathErr) {
-        return callback(realpathErr);
-      }
-
-      if (!globFile.originalSymlinkPath) {
-        // Store the original symlink path before the recursive call
-        // to later rewrite it back.
-        globFile.originalSymlinkPath = globFile.path;
-      }
-      globFile.path = filePath;
-
       // Recurse to get real file stat
-      resolveFile(globFile, enc, callback);
+      fs.stat(globFile.path, onStat);
```

Comment:
> That would really reflect poorly on node fs

^ Remember that I had to create a custom mkdir method because node's `fs` doesn't create with proper permissions.